apps/admin-portal/src/app/resources/overview/resource-types/recipes/models.ts

```diff
@@ -1,4 +1,93 @@
+            itemId: formBuilder.control(value?.itemId ?? '', [Validators.required]),
+            quantity: formBuilder.control(value?.quantity ?? 1, [Validators.required, Validators.min(1)]),
+        });
+
+        formGroup.controls.itemId.valueChanges
+            .pipe(
+                startWith(formGroup.value.itemId),
+                filter((itemId) => itemId !== null && resourceService.resourceId() !== TEMP_RESOURCE_ID),
+                pairwise(),
+                switchMap(([curr, next]) =>
+                    merge(
+                        iif(
+                            () => Boolean(curr),
+                            recipesService[type === 'input' ? 'removeInputFromRecipe' : 'removeOutputFromRecipe'](
+                                resourceService.resourceId(),
+                                curr
+                            ).pipe(
+                                tap(() => {
+                                    const resource = resourceService.resource() as RecipeDto;
+
+                                    const removedRecipeItem = resource[type === 'input' ? 'inputs' : 'outputs'].find(
+                                        ({ itemId }) => itemId === curr
+                                    );
+                                    const indexOf =
+                                        resource[type === 'input' ? 'inputs' : 'outputs'].indexOf(removedRecipeItem);
+
+                                    resource[type === 'input' ? 'inputs' : 'outputs'].splice(indexOf, 1);
+
+                                    resourceService.resource.set(resource);
+                                    resourceService.updatedResource.set(resource);
+                                })
+                            ),
+                            of(curr)
+                        ),
+                        of(formGroup.value.quantity).pipe(
+                            switchMap((quantity) =>
+                                recipesService[type === 'input' ? 'addInputToRecipe' : 'addOutputToRecipe'](
+                                    resourceService.resourceId(),
+                                    serialize(RecipeItemDto, { itemId: next, quantity: quantity })
+                                )
+                            ),
+                            tap((recipeItem) => {
+                                const resource = resourceService.resource() as RecipeDto;
+
+                                resource[type === 'input' ? 'inputs' : 'outputs'].push(recipeItem.toDto());
+                                resourceService.resource.set(resource);
+                                resourceService.updatedResource.set(resource);
+                            })
+                        )
+                    )
+                ),
+                takeUntilDestroyed(destroyRef)
+            )
+            .subscribe();
+
+        return formGroup;
+    });
+}
+
+export function provideControlContainer(): FactoryProvider {
```

Comment:
This provider function is used to make forms modular. This means we can split up large forms into individual components and do different things in those components depending on the state of the form. Eventually, everything will hook back to the encapsulation form.